### ISSUE TITLE
If-Else pattern, "let" element

### ISSUE THREAD

**Issue by niieani** [aurelia/templating-resources#171]
I'm finding myself constantly writing code that seems like an anti-pattern to me:

``` html
<div if.bind="property | transform | something:param1 | pluck:'blah' & signal:'some-signal'">
   (something)
</div>
<div if.bind="!(property | transform | something:param1 | pluck:'blah') & signal:'some-signal'">
   (something else)
</div>
```

This is clearly an If-Else pattern, but here it feels like an anti-pattern, because of:
1. repetition of code (not DRY)
2. double execution of the whole transformation chain

Is there a better way to do this that I'm not aware of? How could an `else` be implemented as a custom attribute? Perhaps something like:

``` html
<template if.bind="property | transform | something:param1 | pluck:'blah' & signal:'some-signal'">
  <true>
    (true content)
  </true>
  <false>
    (false content)
  </false>
</template>
```

Thoughts?

**Comment by EisenbergEffect** [comment]
@martingust @jdanyow I'm wondering if we could build an `else` behavior that could somehow connect to a previous `if` behavior in the DOM. I've had this at the back of my mind. Over the years, I've encountered this scenario on almost every project and it's a pain to repeat positive and negative if blocks. If we do this, we might also  consider elseif and even switch case as well. I don't know what the answer is, but it's worth thinking about. If we do this, we also need to stress the importance of not putting complex logic inside a template....we don't want to solve on problem and create another.

**Comment by jdanyow** [comment]
One thing that works today is:

```html
<div ref="foo" bar.bind="property | transform | something:param1 | pluck:'blah' & signal:'some-signal'">
  <div if.bind="foo.bar">
     (something)
  </div>
  <div if.bind="!foo.bar">
     (something else)
  </div>
</div>
```
This effectively stores the result of the binding expression on an ad-hoc property which can then be referenced in other binding expressions.

---

I think it would be nice if we had a template-controller element called `<let>` that has no UI, all it does is set properties on the `overrideContext` so that they are available for binding purposes:
```html
<let foo.bind="property | transform | something:param1 | pluck:'blah' & signal:'some-signal'" 
     bar.bind="some | other:expression">
</let>

<div if.bind="foo === 'a'">
   (something)
</div>
<div if.bind="foo === 'b'">
   (something)
</div>
<div if.bind="foo === 'c'">
   (something)
</div>

<div if.bind="bar">
  (something)
</div>
<div if.bind="!bar">
  (something)
</div>
```

**Comment by jdanyow** [comment]
Here's an implementation of the `<let>` element.  Turns out it didn't really need to be a template controller...

```js
import {
  customElement,
  noView,
  TargetInstruction,
  inject
} from 'aurelia-framework';

@customElement('let')
@noView()
@inject(TargetInstruction)
export class LetElement {
  instruction = null;
  bindings = null;
  
  constructor(instruction) {
    this.instruction = instruction;
  }
  
  bind(bindingContext, overrideContext) {
    let expressions = this.instruction.expressions;
    let i = expressions.length;
    let bindings = this.bindings = [];
    while (i--) {
      let binding = expressions[i].createBinding(overrideContext);
      binding.bind({ bindingContext, overrideContext });
      bindings.push(binding);
    }
  }
  
  unbind() {
    let i = this.bindings.length;
    while (i--) {
      this.bindings[i].unbind();
    }
    this.bindings = null;
  }
}
```

http://plnkr.co/edit/zGI9d7?p=preview

**Comment by EisenbergEffect** [comment]
This is an interesting idea. I like it because it could enable all the scenarios with a single concept, plus open up some other possibilities. We don't necessarily need to do this as a custom element. We could bake in knowledge of "let" into the compiler. This would make it more efficient. The result of such an element could be to simply parse the attributes into binding expressions that were added to the view factory. Thoughts?

**Comment by jdanyow** [comment]
That would be great, a lot more lightweight, no BehaviorPropertyObservers and could potentially skip parts of the template compilation work when it's a `let` element because we know it shouldn't have any behaviors, content, etc.  Maybe even remove the `let` element from the template so that it doesn't interfere with css, etc.

**Comment by EisenbergEffect** [comment]
Yes, exactly. It would be pretty simple. Here's the place it would go: https://github.com/aurelia/templating/blob/master/src/view-compiler.js#L283

**Comment by jdanyow** [comment]
ok cool! Let me see if I can come up with a PR after I get a couple binding issues fixed.

**Comment by EisenbergEffect** [comment]
Sounds good.

**Comment by niieani** [comment]
I love `let`! Great thinking @jdanyow.
As a side note, it would be great if the docs for `ref` would have this type of usage documented and explained, cause I did read about it, but did not realize you can use it for storing bindings like that. (in fact I was trying to do exactly that, and was bumping my head on the wall, trying to figure out why won't a nested `<template bindable="foo" foo.bind="binding">` work for that same purpose.

**Comment by jdanyow** [comment]
@EisenbergEffect I was looking at the implementation details some more. These bindings differ from standard bindings in a few ways, most importantly, the `target` of these bindings is the `overrideContext`.  This means when `expression.createBinding(target)` is called by the view factory, it's too early and the target argument isn't the right thing / not needed.

I was thinking we should create dedicated expression and binding classes for this feature: `LetExpression` and `Let`. These would be simplified versions of `BindingExpression` and `Binding`. The benefit would be cleaner code for the binding itself and things that work with bindings like binding behaviors and the repeat. Couple downsides to this approach: `<let foo="${bar}"></let>` wouldn't work and we'd probably need to expand the BindingLanguage API [here](https://github.com/aurelia/templating/blob/master/src/view-compiler.js#L304) with an `isLet` argument or something.

Thoughts?

**Comment by EisenbergEffect** [comment]
Can you put together a full proposal? Including proposed syntax, use case examples and what we would need to change to make it happen?

**Comment by ChrisTalman** [comment]
May I ask what the situation is with `else`, `elseif`, and `switch`? These would seem to be very useful for creating cleaner and more comprehensible templates, but would not seem to have been introduced to Aurelia yet.

**Comment by EisenbergEffect** [comment]
@jdanyow @martingust We should revisit this. What do you guys think?

**Comment by cb-fred** [comment]
Just adding my two cents: The let, switch, else-if, are all very nice, but in practice all I ever needed was an else. That problem as stated by OP is very simple and could use a simple solution. Something that can be understood by a designer who may just want to move dom elements around without understanding everything. 

Something like 
```
<div if.bind='foo'>
   ...
</div>
<div else>
  ...
</div>
```

**Comment by mbregante** [comment]
I'm interested in a solution for this kind of scenarios. I found myself writing some _anti-pattern_ as @niieani described and while doing that, thinking "this is not good... hopefully there's a better way to do it". Usually I rework some VM property to make the view more _simple_ but still not ideal.

**Comment by arekbal** [comment]
Here is what I am doing:
`
function contentProcessor(compiler, resources, node, instruction) : boolean
{    
    let trueContent: string = void(0)
    let falseContent: string = void(0)

    for(const n of node.children)
    {
        if(n.nodeName == "TRUE")
        {
            trueContent = n.innerHTML 
            continue        
        }
       
        if(n.nodeName == "FALSE")
        {
            falseContent = n.innerHTML
            continue                 
        }
    }

    if(!trueContent && !falseContent)
    {
        node.innerHTML = `<template replace-part="true">${node.innerHTML}</template>`
        return true
    }

    let output = ""

    if(trueContent)
        output = `<template replace-part="true">${trueContent}</template>`

    if(falseContent)
        output += `<template replace-part="false">${falseContent}</template>`

    node.innerHTML = output
    return true
}

@processContent(contentProcessor)
@customElement("if")
export class DoixIfElement
{
    @bindable value : boolean
}
`
`
<template bindable="value">
    <template if.one-way="value" part="true"></template>
    <template if.one-way="!value" part="false"></template>
</template>
`

Effect:
`
 <if value.bind="true">
        True
 </if>
 <if value.bind="x">
     <true>True</true>
     <false>False</false>
 </if>
`
Still, many cases to consider (same problems as templates replace-part) but I am not going to worry about them currently.

**Comment by Kukks** [comment]
@arekbal Your solution works like a charm!

**Comment by gregorydickson** [comment]
+1 on if/else

**Comment by Kukks** [comment]
@arekbal After using it for a while, it seems like IE did not like it(out of stack space errors during loading time) and I had to revert back to previous bindings :(

**Comment by weoreference** [comment]
Hello guys, how are you?

I am wondering if the Aurelia team plans to officially support this feature?
I've been running into it since day 1 -- even back in Durandal. Just seems to be a general problem in conditional layouts.

Anyway, thank you.

**Comment by arekbal** [comment]
@Kukks 
https://github.com/aurelia/dialog/issues/45
It seems to be related... just like I said... same problems as templates.

Works with Edge, and is out of stack space on IE. <template>s fail in _compile
Template elements are not supported by IE and are showing as HTMLUnknownElement.

I went throught recurrenting aurelia code in IE debugger - node processing hits nodeType 1 instead of 11 
and later on there is no source.content both of which are part of HTMLTemplateElement spec. It could probably be hacked/fixed(in aurelia or in customElement) but I had no time to do so. It seems a bit more time consuming than I initially thought.

**Comment by mgthomas99** [comment]
I recently wrote an [npm package](https://www.npmjs.com/package/aurelia-conditionals) which addresses the if/else issue using HTML `<slot>` elements.

**Comment by nashwaan** [comment]
It looks like this is now released as part of core aurelia as mentioned in [Aurelia Release Notes - Early October 2017](http://blog.aurelia.io/2017/10/03/aurelia-release-notes-early-october-2017/):

> If/Else - We've added an else attribute that you can use in combination with the if attribute in your templates! No more need for inverse if statements to handle the falsy case.`

I couldn't find any documentation about it. Is it same as proposed by @cb-fred ?

```html
<div if.bind='foo'>
   ...
</div>
<div else>
  ...
</div>
```

**Comment by EisenbergEffect** [comment]
That's how it works, yes.

**Comment by dkent600** [comment]
This is great, would've been nice if it also worked with 'show'  (show/else)

**Comment by RomkeVdMeulen** [comment]
Any plans for `switch` or `else if`?

**Comment by soulhunter1987** [comment]
One more vote for switch support!

**Comment by MJSanfelippo** [comment]
Any thoughts on else if? Run into a few use cases for it.

**Comment by bigopon** [comment]
@MJSanfelippo Can you help put your use cases into a feature request issue? Also would be great if you could give some more thought/explanation for it 👍

**Comment by MJSanfelippo** [comment]
@bigopon Can't post company code, but can give a vague rundown. We have a list of things and each thing has a property that's an enum. There are 15-20 different values this enum could be. We're conditionally binding a certain child component based on the value of that enum. So that's 15-20 times it needs to check the if.bind whereas if we had if, else if, else if, else if... else if, else, it would only check until it hit the correct one (just like a normal else, else if, else block in programming). Obviously checking the value of an enum is not a huge performance hit, but this specific component that conditionally binds based on the enum is on the page as many times as the user sets it to which could be anywhere between 10-100+, we don't know really know.

Basically, the use case is the same use case that using if, else if, else in a programming language is. It's inefficient to just do if (thing) if (otherThing) etc.
